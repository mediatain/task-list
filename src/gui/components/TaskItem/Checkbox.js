import React from 'react';
import styled from 'styled-components';
import { disabledColor } from '../../colors';

const StyledCheckbox = styled.div`
  display: flex;
  align-items: center;
  padding-right: 1.7rem;
  padding-left: .1rem;
  cursor: pointer;
`;

const Box = styled.div`
  position: relative;
  border: .1rem solid;
  border-radius: .25rem;
  height: 1.5rem;
  width: 1.5rem;
  color: ${disabledColor};
`;

const Mark = styled.div`
  position: absolute;
  left: .3rem;
  top: .5rem;
  width: .2rem;
  height: 1rem;
  background: ${disabledColor};
  transform: rotate(315deg);
  display: ${({ checked }) => checked ? 'block' : 'none'};

  &::after {
    position: absolute;
    left: 100%;
    bottom: 0;
    content: "";
    background: inherit;
    width: 2rem;
    height: .2rem;
  }
`;

export default function Checkbox({
  checked = false,
  onClick,
}) {
  return (
    <StyledCheckbox onClick={onClick}>
      <Box>
        <Mark checked={checked} />
      </Box>
    </StyledCheckbox>
  );
}

