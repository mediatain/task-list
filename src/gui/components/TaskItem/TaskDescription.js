import styled from 'styled-components';

export default styled.span`
  font-size: 1.5rem;
  padding: 1rem;
`;
