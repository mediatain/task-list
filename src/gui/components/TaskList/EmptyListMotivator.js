import styled from 'styled-components';
import { textColorDark } from '../../colors';

export default styled.div`
  position: relative;
  font-size: 2rem;
  color: ${textColorDark};
  padding-left: 5rem;
  padding-top: 2rem;

  &::before {
    position: absolute;
    left: 1rem;
    top: -.5rem;
    content: "â¤¶";
    font-size: 200%;
    font-weight: 200;
    transform: rotate(90deg);
  }
`;
