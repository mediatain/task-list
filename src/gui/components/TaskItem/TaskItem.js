import React from 'react';
import CheckBox from './Checkbox';
import TaskDescription from './TaskDescription';
import styled, { css } from 'styled-components';
import { disabledColor, textColorDark } from '../../colors';

const StyledTaskItem = styled.li`
  display: flex;
  flex-direction: row;
  list-style: none;
  color: ${textColorDark};
  
  ${({ done }) => done && css`
    text-decoration: line-through;
    color: ${disabledColor};
  `}
`;

export default function TaskItem({
  task,
  onCheck,
}) {
  return (
    <StyledTaskItem done={task.done}>
      <CheckBox checked={task.done} onClick={onCheck} />
      <TaskDescription>{task.description}</TaskDescription>
    </StyledTaskItem>
  );
}

