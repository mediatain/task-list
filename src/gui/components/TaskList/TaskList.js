import React from 'react';
import TaskItem from '../TaskItem/TaskItem';
import EmptyListMotivator from './EmptyListMotivator';
import styled from 'styled-components';

const StyledTaskList = styled.ul`
  max-width: 40rem;
  margin: 2rem auto;
  padding: 0;
`;

export default function TaskList({
  tasks = [],
  onToggleTask = () => {},
}) {
  return (
    <StyledTaskList>
      {tasks.map(task => (
        <TaskItem
          key={task.id}
          task={task}
          onCheck={() => onToggleTask(task.id)}
        />
      ))}
      {tasks.length === 0 && (
        <EmptyListMotivator>Add your first task!</EmptyListMotivator>
      )}
    </StyledTaskList>
  )
}
