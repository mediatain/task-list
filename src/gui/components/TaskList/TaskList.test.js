import React from 'react';
import 'jest-styled-components';
import { render } from '@testing-library/react';
import { shallow } from 'enzyme';
import TaskList from './TaskList';
import EmptyListMotivator from './EmptyListMotivator';
import TaskItem from '../TaskItem/TaskItem';

jest.mock('../TaskItem/TaskItem', () => jest.fn().mockReturnValue('TaskItem'));

describe('TaskList', () => {
  const cases = [
    [[]],
    [[{ id: 'Task 1' }]],
    [[{ id: 'Task 1' }, { id: 'Task 2' }, { id: 'Task 3' }]],
  ];

  it.each(cases)('should match the snapshot for tasks %p', (tasks) => {
    const { container } = render(<TaskList tasks={tasks} />);

    expect(container).toMatchSnapshot();
  });

  it.each(cases)('should render one TaskItem for each task', (tasks) => {
    const taskList = shallow(<TaskList tasks={tasks} />);

    expect(taskList.find(TaskItem).length).toEqual(tasks.length);
  });

  it('should render an EmptyListMotivator if tasks are empty', () => {
    const taskList = shallow(<TaskList tasks={[]} />);

    expect(taskList.exists(EmptyListMotivator)).toBe(true);
  });

  it('should render an EmptyListMotivator if tasks are undefined', () => {
    const taskList = shallow(<TaskList />);

    expect(taskList.exists(EmptyListMotivator)).toBe(true);
  });

  it('should not render an EmptyListMotivator if there are tasks', () => {
    const taskList = shallow(<TaskList tasks={[{ id: 'Task 1' }, { id: 'Task 2' }]} />);

    expect(taskList.exists(EmptyListMotivator)).toBe(false);
  });

  describe('when the TaskItem gets checked', () => {
    it('should call the passed onToggleTask function with the task id', () => {
      const onToggleTask = jest.fn();
      const taskId = 'Task 1';
      const taskList = shallow(<TaskList tasks={[{ id: taskId }]} onToggleTask={onToggleTask} />);

      taskList.find(TaskItem).simulate('check');

      expect(onToggleTask).toBeCalledWith(taskId);
    });
  });
});

