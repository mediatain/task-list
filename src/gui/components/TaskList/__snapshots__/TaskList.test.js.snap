// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList should match the snapshot for tasks [[Object], [Object], [Object]] 1`] = `
.c0 {
  max-width: 40rem;
  margin: 2rem auto;
  padding: 0;
}

<div>
  <ul
    class="c0"
  >
    TaskItem
    TaskItem
    TaskItem
  </ul>
</div>
`;

exports[`TaskList should match the snapshot for tasks [[Object]] 1`] = `
.c0 {
  max-width: 40rem;
  margin: 2rem auto;
  padding: 0;
}

<div>
  <ul
    class="c0"
  >
    TaskItem
  </ul>
</div>
`;

exports[`TaskList should match the snapshot for tasks [] 1`] = `
.c1 {
  position: relative;
  font-size: 2rem;
  color: #000000;
  padding-left: 5rem;
  padding-top: 2rem;
}

.c1::before {
  position: absolute;
  left: 1rem;
  top: -.5rem;
  content: "â¤¶";
  font-size: 200%;
  font-weight: 200;
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

.c0 {
  max-width: 40rem;
  margin: 2rem auto;
  padding: 0;
}

<div>
  <ul
    class="c0"
  >
    <div
      class="c1"
    >
      Add your first task!
    </div>
  </ul>
</div>
`;
